vision/datasets/imageloader.go

```diff
@@ -4,99 +4,117 @@ import (
 	"image"
 	"io"
 	"path/filepath"
+	"unsafe"
 
 	torch "github.com/wangkuiyi/gotorch"
 	tgz "github.com/wangkuiyi/gotorch/tool/tgz"
 	"github.com/wangkuiyi/gotorch/vision/transforms"
 )
 
-// ImageLoader struct
-type ImageLoader struct {
-	r      *tgz.Reader
-	vocab  map[string]int64
-	err    error
-	inputs []torch.Tensor
-	labels []int64
-	trans  *transforms.ComposeTransformer
-	mbSize int
+type sample struct {
+	img   transforms.ImageFloat
+	label int
 }
 
-func (p *ImageLoader) tensorGC() {
-	p.inputs = []torch.Tensor{}
-	p.labels = []int64{}
-	torch.GC()
+// ImageLoader struct
+type ImageLoader struct {
+	r       *tgz.Reader
+	vocab   map[string]int
+	samples chan sample
+	err     chan error
+	trans   *transforms.ComposeTransformer
+	mbSize  int
 }
 
 // NewImageLoader returns an ImageLoader
-func NewImageLoader(fn string, vocab map[string]int64, trans *transforms.ComposeTransformer, mbSize int) (*ImageLoader, error) {
+func NewImageLoader(fn string, vocab map[string]int, trans *transforms.ComposeTransformer, mbSize int) (*ImageLoader, error) {
 	r, e := tgz.OpenFile(fn)
 	if e != nil {
 		return nil, e
 	}
-	return &ImageLoader{
-		r:      r,
-		vocab:  vocab,
-		err:    nil,
-		trans:  trans,
-		mbSize: mbSize,
-	}, nil
+	m := &ImageLoader{
+		r:       r,
+		vocab:   vocab,
+		samples: make(chan sample, mbSize*4),
+		err:     make(chan error),
+		trans:   trans,
+		mbSize:  mbSize,
+	}
+	go m.retreiveMinibatch()
+	return m, nil
 }
 
 // Scan return false if no more dat
 func (p *ImageLoader) Scan() bool {
-	if p.err != nil {
-		return false
+	select {
+	case e := <-p.err:
+		if e != nil {
+			return false
+		}
+	default:
+		return true
 	}
-	p.tensorGC()
-	p.retreiveMinibatch()
-	return p.err == nil || p.err == io.EOF // the next call will return false
+	return true
 }
 
 func (p *ImageLoader) retreiveMinibatch() {
+	defer close(p.samples)
+	defer close(p.err)
 	for {
 		hdr, err := p.r.Next()
 		if err != nil {
-			p.err = err
+			p.err <- err
 			break
 		}
 		if !hdr.FileInfo().Mode().IsRegular() {
 			continue
 		}
 		classStr := filepath.Base(filepath.Dir(hdr.Name))
 		label := p.vocab[classStr]
-		p.labels = append(p.labels, label)
 
 		m, _, err := image.Decode(p.r)
 		if err != nil {
-			p.err = err
+			p.err <- err
 			break
 		}
 		input := p.trans.Run(m)
```

Comment:
There are tensor operations in transforms. So, when the tensors are freed?

Reply:
I added a function `splitComposeByToTensor` to splite the composeTransoform into `trans1` and `trans2` by  `ToTensorTransform`. `trans1` running in the goroutine, and `trans2` running at retrieving the minibatch. So we can free the Tensors at each iteration.